### DISCUSSION TITLE
How can I turn username --> email address using LDAP?

### DISCUSSION THREAD

**Discussion by snaildos** [docker-mailserver/docker-mailserver#4085] [Q&A]
I've read the documentation but can only see use cases when the email address the LDAP user uses is the domain that docker-mailserver hosts.
I however, want to be able to have a username provision to username@example.com using docker-mailserve.r
Any ideas?
Warm Regards,
Snail

**Comment by polarathene** [comment]
Postfix presently appends `@example.com` to the SASL login username when no domain-part (`@some-value`) is provided. `@example.com` would be used when DMS has the `compose.yaml` setting `hostname: mail.example.com`. That would only apply for Postfix ports 587 and 465.

This is due to the Postfix `main.cf` setting using the [`mydomain` setting](https://www.postfix.org/postconf.5.html#mydomain):

https://github.com/docker-mailserver/docker-mailserver/blob/e370c0c96a28acfbd66b20ffac327528eda78104/target/postfix/main.cf#L99

I can't quite recall specifics with LDAP involved, Postfix would query LDAP with the username I think and if a result was returned, use the `mail` attribute value:

https://github.com/docker-mailserver/docker-mailserver/blob/e370c0c96a28acfbd66b20ffac327528eda78104/target/postfix/ldap-users.cf#L5

I think this had the query use the login username to query the mail address for that user lookup:

https://github.com/docker-mailserver/docker-mailserver/blob/e370c0c96a28acfbd66b20ffac327528eda78104/target/postfix/ldap-users.cf#L4

So you may need to alter that via the current ENV support. I do recall that SASLAuthd complicates that a little further, and presently has a configuration that makes that a little bit inconsistent, but many users mistakenly enable SASLAuthd when you don't actually need that to use LDAP. So if you have that enabled, try without it.

On the Dovecot side for other ports that need your LDAP user, this is configured here:

https://github.com/docker-mailserver/docker-mailserver/blob/e370c0c96a28acfbd66b20ffac327528eda78104/target/dovecot/dovecot-ldap.conf.ext#L10-L11

You'd currently use the ENV to adjust that. Dovecot has separate syntax for querying with only the login username, or also including a domain-part, which in the example above is [`%n`](https://doc.dovecot.org/configuration_manual/config_file/config_variables/#config-variables) (_just the local-part, anything before an `@`_), while `%u` would be the full email address. Make sure you query on the correct attribute.

---

There is a PR I need to finish that refactors the LDAP support, along with improved LDAP docs to support that, but I've not had time to focus on that for a while: https://github.com/docker-mailserver/docker-mailserver/pull/3524

**Reply by snaildos** [reply]
Hey, thanks for the reply.
So overall, I need to modify Postfix/main.cf, ldap-users.cf, and add result & query.
How do I know if I need SASLAuthd?
Then I also add Dovecot too ( user_filter = (&(objectClass=PostfixBookMailAccount)(cn=%n)) ).
What should I do for user_attrs? I don't want to add additional attributes, can I just hard set a response such as mailHomeDirectory, mailUid and gId as well as mailStorage?
Still just a little confused about overall config as I want to do it the easiest way possible without adding extra attributes, etc.
Kind regards,
Snail

**Reply by polarathene** [reply]
> So overall, I need to modify Postfix/main.cf, ldap-users.cf, and add result & query.

If you need to actually modify `main.cf` use our override file `postfix-main.cf` support, try not to manually override `/etc/postfix/main.cf` (especially avoid mounting a file directly to that). If for some reason that isn't viable, you could try our `user-patches.sh` feature in the docs to run `postconf key=value` to update settings.

`ldap-users.cf` likewise AFAIK shouldn't be adjusted directly. Presently we offer the ENV approach as should be documented for overriding any values configured in that file.

---

> How do I know if I need SASLAuthd?

If you don't know, then you very likely don't need it. It was contributed by a community member as an alternative method for SASL support. By default Postfix delegates to Dovecot for this and that's usually all you need. Dovecot and Postfix have their own LDAP integration, so the additional abstraction through SASLAuthd isn't really necessary.

---

> Then I also add Dovecot too ( `user_filter = (&(objectClass=PostfixBookMailAccount)(cn=%n))` ).

Again, the LDAP docs should have ENV based configuration example for this. I know the current state of the docs for LDAP are not ideal, the LDAP support is community maintained/contributed as no maintainer uses it. I was more familiar with it last year but my recall is a bit foggy.

---

> What should I do for user_attrs? I don't want to add additional attributes, can I just hard set a response such as mailHomeDirectory, mailUid and gId as well as mailStorage?

I'm not able to provide much assistance here. I barely recall the specifics with LDAP and the integration we have. I will get around to refreshing on that to finish the referenced PR and updated docs, but I can't say how long that will take until I find time for it.

In the meantime, our current test suite coverage for LDAP does have the following that I documented for context, it may help?:

https://github.com/docker-mailserver/docker-mailserver/blob/e370c0c96a28acfbd66b20ffac327528eda78104/test/tests/serial/mail_with_ldap.bats#L37-L121

The related LDAP config files associated to supporting that test can be found [here](https://github.com/docker-mailserver/docker-mailserver/tree/e370c0c96a28acfbd66b20ffac327528eda78104/test/config/ldap). Where it seems that files like `ldap-users.cf` are added to `/tmp/docker-mailserver` (config volume) as overrides, so you can perhaps try that:

https://github.com/docker-mailserver/docker-mailserver/blob/e370c0c96a28acfbd66b20ffac327528eda78104/test/tests/serial/mail_with_ldap.bats#L164-L165

---

> Still just a little confused about overall config as I want to do it the easiest way possible without adding extra attributes, etc.

I understand, it's just a lack of time on my part to assist with that. The LDAP support and docs don't see much community contributions for some time, you might want to see if the LDAP support is better for alternatives to DMS?

**Reply by snaildos** [reply]
Hey, thanks for the reply.
Is it possible to talk further on a live chat platform such as Telegram, Matrix or Discord?
I have a few more questions to attempt to get this working, and I don't know anyone else I could ask regarding LDAP integration on docker-mailserver.
Thanks,
Snail

**Reply by polarathene** [reply]
> Is it possible to talk further on a live chat platform such as Telegram, Matrix or Discord?
> I have a few more questions to attempt to get this working, and I don't know anyone else I could ask regarding LDAP integration on docker-mailserver.

No sorry. As you can see I'm a bit limited with time to spare, so it wouldn't change my response time 😓 

It's also better for other LDAP users if we discuss here, since they can discover this via search and also find information to assist them.

**Reply by kraghuprasad** [reply]
> `/tmp/docker-mailserver`

This folder is not getting used, if I enable LDAP.  No configuration files are available in this folder when I set various LDAP related parameters in `mailserver.env` file.

Due to this, we won't be able to even modify the files like `/etc/postfix/ldap-*.cf` manually, even if auto-generation of LDAP configuration turns out to be unreliable.

**Reply by polarathene** [reply]
> This folder is not getting used, if I enable LDAP. No configuration files are available in this folder when I set various LDAP related parameters in `mailserver.env` file.

You misunderstood. `/tmp/docker-mailserver` allows you to provide these config files yourself, and they will be copied over to the internal config location. We already provide these files by default, so you don't have to provide your own, the ENV you set will modify the internal copies:

https://github.com/docker-mailserver/docker-mailserver/blob/e370c0c96a28acfbd66b20ffac327528eda78104/test/tests/serial/mail_with_ldap.bats#L212-L216

> Due to this, we won't be able to even modify the files like /etc/postfix/ldap-*.cf manually, even if auto-generation of LDAP configuration turns out to be unreliable.

You modify those via providing your own configs to `/tmp/docker-mailserver` and/or using ENV.

**Reply by kraghuprasad** [reply]
OK, got it. Thanks for the clarifications. The confusion arose because when I use file based user-accounts, this folder had many configuration files. But as soon as I enabled LDAP, this folder was empty.

I am trying to debug why I am not able to login to an existing mail account kept in LDAP. If I want to change configurations in a running container, would it suffice to make changes in /etc/postfix/* and /etc/dovecot/* and test the setup? Or are we getting configurations from some other locations?

**Reply by polarathene** [reply]
> I am trying to debug why I am not able to login to an existing mail account kept in LDAP

Ah, there are some gotchas IIRC with the LDAP implementation. If you have enabled SASLAUTHD try to disable that, you shouldn't actually need it, it was contributed by the same contributor of LDAP support for their own needs (_without dovecot_). Postfix and Dovecot have some related config (_which I seem to have [already documented above](https://github.com/orgs/docker-mailserver/discussions/4085#discussioncomment-9854562) for reference_).

> If I want to change configurations in a running container, would it suffice to make changes in `/etc/postfix/*` and `/etc/dovecot/*` and test the setup?

Within the container you could make those changes and use `postfix reload` / `dovecot reload` commands, if those don't have effect, you can also use `supervisorctl restart postfix` / `supervisorctl restart dovecot`.

Just note if you are directly mounting files via volumes that changes will not update in the container if the inode of the file on the Docker host changes (_this happens depending on how your text editor saves changes to a file_), the only way to avoid that is directory mounts instead but that will replace the existing contents of the directory within the container which probably won't work here. I tend to use `nano` command in the container to manage files directly, or you can mount a folder somewhere else and use `cp /from/path/file /to/path/file` to update the real internal configuration.

DMS does have a change detection service that detects config files in the Config Volume (mounted to `/tmp/docker-mailserver`) but presently it's disabled when LDAP is enabled, so you can't leverage that.

**Reply by kraghuprasad** [reply]
Thanks for the info.

I am planning to do directory-mount in docker-compose.yml to /tmp/docker-mailserver and use my own *.cf files.  I wanted to change configuration files in the container only for the purpose of debugging. At present I think, I have misconfigured ENVs in my docker-compose file, due to which I am unable to authenticate via my LDAP server while sending out emails to external parties.

**Comment by snaildos** [comment]
Hey @polarathene ,
Thank you for all the feedback you've given.
I've gotten everything working from the configuration you've given to me aside from sending.
When sending an email I get:
"Sender address rejected: not owned by user USERNAME;"
I have almost identical configuration to the testing documents you've given to me, aside from "DOVECOT_PASS_ATTRS" (I've removed) an additionally using `LDAP_QUERY_FILTER_SENDERS=(|(mail=%s)(mail=admin@*))` (The user I am sending emails from is called 'test'). 
I hope to hear some insight from you. Thanks for your help prior.
Warm regards,
Snail

**Reply by polarathene** [reply]
> I've gotten everything working from the configuration you've given to me aside from sending.

You might need to search our issue history (and maybe discussion history) here to try find previous LDAP issues/discussions. They'll likely have advice you're seeking on any issues you run into.

Here's a recent issue related to LDAP sending for example: https://github.com/docker-mailserver/docker-mailserver/issues/4029

---

> When sending an email I get:
> "Sender address rejected: not owned by user USERNAME;"

This suggests you have `SPOOF_PROTECTION=1`. That will restrict you to only being able to send mail with a sender address allowed for your user login name. I believe we presently have our LDAP `SPOOF_PROTECTION=1` tests disabled (or at least partially) because of a bug there. I did investigate that over a year ago and it was due to the default LDAP configs IIRC.

From memory, when you connect to Postfix (ports 587 / 465) with your mail client to submit mail to send. You'll authenticate with a username, this may look like `john.doe` or `john.doe@example.com`. If it's just `john.doe` then Postfix will add `@example.com` for the SASL login I think, and that may be part of the LDAP lookup query for `ldap-users.cf`, while the table for `SPOOF_PROTECTION=1` is separate, which you seem to be aware of with the ENV `LDAP_QUERY_FILTER_SENDERS` 👍 

The way this check works I think was the sender address being used would be queried, LDAP receives that query, and then the result attribute is used for any returned responses to filter which login usernames are permitted to send with that sender address. Here's the defaults we currently have:

https://github.com/docker-mailserver/docker-mailserver/blob/e370c0c96a28acfbd66b20ffac327528eda78104/target/postfix/ldap-senders.cf#L4-L5

So typically that's returning `mail` (`john.doe@example.com`) and `uid` (the User ID `john.doe`). Your LDAP values for those attributes may not be the same though. I can't quite recall the specifics for that bug, but do recall an issue with SASLAuthD feature (_it would not remove the `@example.com` "realm" from Postfix, unless config was updated_).

If you're not using SASLAuthD anymore, you shouldn't have an issue there. I could be mistaken, but Postfix should be trying to match either `john.doe` or `john.doe@example.com` formats for your "login username" (_without LDAP this is what is provided to `setup email add`, which treats the email as the DMS account username that Postfix and Dovecot use_). You need to make sure the LDAP query is working correctly, and then it's just making sure that Postfix can match the given client user login name with the attribute response returned from LDAP. After that it should work as expected.

Alternatively don't use `SPOOF_PROECTION=1` if that's acceptable. This feature is only to restrict what sender address you can use to send mail with. Without it, you could attempt to send mail with domains you don't own (`john.doe@github.com` / `john.doe@gmail.com`), which would fail but DMS wouldn't know any better. It's also used for deployments where you have other users that shouldn't be able to send as others with your mail domain (`eg: john.doe` sending as `ceo@example.com`).

**Reply by snaildos** [reply]
Hello, thank you for your insight. I've gotten everything working as expected from your help - I appreciate your efforts while being low on time.
Just a quick final question, if SPOOF PROTECTION is off how will i know my custom spoof protection is working? It should just deny from sending other emails, correct? Thanks.
Thanks again for your help.
Kind regards,
Snail

**Reply by polarathene** [reply]
> Just a quick final question, if SPOOF PROTECTION is off how will i know my custom spoof protection is working? It should just deny from sending other emails, correct?

`SPOOF_PROTECTION=1` should prevent you from sending mail with any _sender address_ that your account/user has not been allowed to use.

Here is the LDAP tests related to that feature:

https://github.com/docker-mailserver/docker-mailserver/blob/32c0a346fcec975ba133443d98e3a1e7c0d78714/test/tests/serial/mail_with_ldap.bats#L325-L367

The custom `ldap-senders.cf` test at the end there is skipped, I did identify a fix to enable it again but delayed resolving it until after the LDAP refactor PR lands.

I haven't looked at the LDAP stuff for a while, but here is the related portion for configuring a custom `SPOOF_PROTECTION=1` query I documented inline a while ago:

https://github.com/docker-mailserver/docker-mailserver/blob/32c0a346fcec975ba133443d98e3a1e7c0d78714/test/tests/serial/mail_with_ldap.bats#L62-L70

For LDAP with `SPOOF_PROTECTION=1` there is presently two variants configured, when you configure a custom query like shown above, it would only use that query, not sourcing from the users/aliases/groups LDAP sources (_which I think was a bad idea and will probably be dropped in future_):

https://github.com/docker-mailserver/docker-mailserver/blob/32c0a346fcec975ba133443d98e3a1e7c0d78714/target/scripts/startup/setup.d/security/spoofing.sh#L7-L12

---

So if you want to verify the feature is working as intended. Just send mail with _sender address_ that should be allowed, vs one that is not allowed for that account. Maybe also double check by sending from another user account just to verify that a shared _sender address_ works, and that _sender address_ that shouldn't work for that user only works for the account it should 😅

**Reply by snaildos** [reply]
Hello, thank you for the reply.
Everything is working well except sending outbound emails from docker-mailserver.
When sending an email to my @hotmail.com account I get "User doesn't exist:", but internal sending works (to other email addresses in the email server)
Do you have an idea on how to solve this?
Thank you again for all your previous insight regarding the other issues I've had, everything works flawlessly except for this one small issue.
Warm regards

**Reply by polarathene** [reply]
> Do you have an idea on how to solve this?

You likely have an alias or account assigned with an `@hotmail.com` mail domain, this will tell DMS that it owns and manages mail for that mail domain, so it won't attempt to send mail outbound to another mail server. As such it tries to deliver the mail but realizes there is no account for that username, hence the "User doesn't exist".

If you don't have an account or alias with that, only the alias for a local mail domain pointing to an external one like `@hotmail.com`, then this may be a bug with LDAP (or rather the query filter). I believe this is the `QUERY_DOMAIN` like shown here:

https://github.com/docker-mailserver/docker-mailserver/blob/32c0a346fcec975ba133443d98e3a1e7c0d78714/test/tests/serial/mail_with_ldap.bats#L46-L53

https://github.com/docker-mailserver/docker-mailserver/blob/32c0a346fcec975ba133443d98e3a1e7c0d78714/test/tests/serial/mail_with_ldap.bats#L117

The `mailAlias=*@%s` allows for any `mailAlias` with the domain match to be considered as managed by DMS, which it should be AFAIK. If you have that or the similar `mail` attribute returning with `@hotmail.com` address, that is your problem I think.
